test/test_multidevice_pipeline.cpp

```diff
@@ -304,6 +304,135 @@ INSTANTIATE_TEST_SUITE_P(
         ::testing::Values(true),
         ::testing::Bool()));
 
+enum class SchedulingMode {
+  noIntraDeviceScheduling,
+  manualScheduling,
+  reductionScheduler,
+  automaticScheduling,
+};
+
+std::ostream& operator<<(std::ostream& out, const SchedulingMode& mode) {
+  switch (mode) {
+    case SchedulingMode::noIntraDeviceScheduling:
+      return out << "noIntraDeviceScheduling";
+    case SchedulingMode::manualScheduling:
+      return out << "manualScheduling";
+    case SchedulingMode::reductionScheduler:
+      return out << "reductionScheduler";
+    case SchedulingMode::automaticScheduling:
+      return out << "automaticScheduling";
+    default:
+      NVF_ERROR(false);
+  }
+  return out;
+}
+
+using PipelineTestStagedReductionParams = std::tuple<SchedulingMode>;
+class PipelineTestStagedReduction
+    : public PipelineTest,
+      public ::testing::WithParamInterface<PipelineTestStagedReductionParams> {
+};
+
+// 1D staged reduction
+// Inputs: X[A,B,C]
+TEST_P(PipelineTestStagedReduction, staged_reduction) {
+  auto [scheduling_mode] = GetParam();
+
+  int num_devices = communicator->size();
+  int A = num_devices;
+  int B = 8;
+  int C = 32;
+  std::vector<int64_t> unsharded_input_sizes = {A, B, C};
+  std::vector<int64_t> input_sizes(unsharded_input_sizes);
+  input_sizes[0] = 1;
+
+  FusionGuard fg(fusion.get());
+  TensorView* tv0 = makeConcreteTensor(unsharded_input_sizes);
+  TensorView* tv1 = sum(tv0, {2});
+  TensorView* tv_out = sum(tv1, {0});
+  fusion->addInput(tv0);
+  fusion->addOutput(tv_out);
+
+  // multi device scheduling:
+  std::vector<int64_t> devices(num_devices);
+  std::iota(devices.begin(), devices.end(), 0);
+  DeviceMesh mesh(devices);
+  for (auto tv : {tv0, tv1, tv_out}) {
+    tv->setDeviceMesh(mesh);
+  }
+  for (auto tv : {tv0, tv1}) {
+    tv->axis(0)->parallelize(ParallelType::DIDx);
+  }
+
+  // Intra-device reduction scheduling for the first reduction:
+  switch (scheduling_mode) {
+    case SchedulingMode::noIntraDeviceScheduling:
+      break;
+    case SchedulingMode::automaticScheduling:
+      multi_device_executor_params.use_fusion_executor_cache = true;
+      break;
+    case SchedulingMode::reductionScheduler: {
+      auto reduction_params = getReductionHeuristics(
+          fusion.get(), {at::empty(input_sizes, tensor_options)});
+      NVF_CHECK(reduction_params, "Reduction schedule was not generated!");
+      l_params = reduction_params->lparams;
+      scheduleReduction(fusion.get(), *reduction_params);
+      break;
+    }
+    case SchedulingMode::manualScheduling: {
+      // clang-format off
```

Comment:
If it's just about clang-format messing up the comments, put (multiple) `clang-format off` and `on` around the comments, so other real code is still checked.

Reply:
ok!
https://github.com/NVIDIA/Fuser/pull/1925/files#r1523870318